feat: add RabbitMQ publisher

anandsunderraman opened the pull request:
**Description**

Creating a amqp publisher from asyncapi file

**Related issue(s)**
https://github.com/asyncapi/go-watermill-template/issues/132

**Summary of Changes**
- Added functionality for `amqp` publishers under `components/Handlers.js`
- Added functionality to create `amqpPublisher` under `components/Publisher.js`
- Added util functions under `components/common.js`
- Added a util to convert any model to watermill message under `template/asyncapi/payload.js`
- Refactored code to produce the `main.go` file in `template/index.js`
- Added a bunch of unit tests

smoya commented on components/Handlers.js:
```diff
@@ -15,28 +15,81 @@ func ${operation}(msg *message.Message) error {
+const amqpPublisherFunction = (channelName, operation, message) => `
+// ${operation} publish handler for ${channelName}.
+func ${operation}(ctx context.Context, a *amqp.Publisher, payload ${message}) error {
+
```
Nitpick. Remove extra line
```suggestion
```

smoya commented on components/Handlers.js:
```diff
@@ -15,28 +15,81 @@ func ${operation}(msg *message.Message) error {
+}
+
+const amqpPublisherFunction = (channelName, operation, message) => `
+// ${operation} publish handler for ${channelName}.
```
```suggestion
// ${operation} is the publish handler for ${channelName}.
```

dxps commented:
@anandsunderraman Could you please react on the provided feedback?
Meanwhile, I will play with this new part as well.
Thanks!

anandsunderraman commented:
> @anandsunderraman Could you please react on the provided feedback? Meanwhile, I will play with this new part as well. Thanks!

@dxps I was waiting for all reviewers to provide feedback, so that I can make my changes all at once, preventing multiple back and forth

dxps commented:
@anandsunderraman Thanks for the quick feedback!
I already started playing with this useful generator, and I'd like to cover also the publishing part.

So, in order to help (test & evaluate for now), while being new on such generator logic, how could I use it locally (a clone of this repo and on this `rabbit-publisher` branch) :question: 

Thanks.

dxps commented:
@anandsunderraman  While just testing it with the spec mentioned [here](https://github.com/asyncapi/go-watermill-template/issues/138#issuecomment-1079708764), which the spec does not yet declare any `subscribe` operation at the channel level, there are two (small) issues:
1. In `handlers.go`, `"context"` is imported but not used.
2. In `router.go`, there is no `GetAMQPPublisher` generated function, thus in `main.go` there is the corresponding error.

Used `ag --debug /path/to/spec.yaml https://github.com/anandsunderraman/go-watermill-template/tree/rabbit-publisher -o gen_go` for testing it.

dxps commented on components/Handlers.js:
```diff
@@ -15,28 +15,81 @@ func ${operation}(msg *message.Message) error {
   return Object.entries(channels)
     .map(([channelName, channel]) => {
       if (channel.hasPublish()) {
         const operation = pascalCase(channel.publish().id());
```
Exactly! Either:
- some method (handler function) name is auto generated (and have it documented)
- or the docs should explicit that `operationId` should be used to properly define your handler name

dxps replied:
And there is also `$id` property of `payload`, based on which the proper Go `struct` naming is done.
That's another useful and optional thing (without it you get a name, but it's something like `AnonymousSchema` which is not that not that great. It's ok that it is generated, no syntax errors, but the naming can be improved).

anandsunderraman replied:
@smoya @dxps thanks for noticing this.
I am tempted to generate a method name based if `operationId` is not present.
But as @dxps mentions we would have to make sure there are no duplicate names.
Another thing I notice is that channel names can have special characters. So we would have to remove all special characters.
I like @dxps's suggestion to mandate an `operationId` for this spec. Meanwhile I can also check other generators on how they handle this scenario

smoya replied:
@anandsunderraman I reviewed the code again and I noticed you didn't add a a guard checking if the `operationId` is present, forcing the user to set it or throwing an error otherwise. You have an example in [nodejs-ws-template](https://github.com/asyncapi/nodejs-ws-template/blob/c50a34556de766098e7b3b25d95fc064235e3728/filters/all.js#L21-L28).

Also, I would mention this in the README file as for example [java-spring-template](https://github.com/asyncapi/java-spring-template/blob/7feab3c2d5b2800828e25e28c30eb0deed56e90d/README.md#asyncapi-definitions) does.

dxps replied:
Good catch! Forgot about `operationId`, although I mentioned it initially.

The other part that could get improved is related to the `$id` property of payload, based on which the proper Go struct naming is done. That's another useful and optional thing, without specifying it you get a name like `AnonymousSchema` which is not that great or useful.

dxps commented on template/index.js:
```diff
@@ -1,90 +1,161 @@
+        const msgName = pubConfig.message.toLowerCase();
+        return `
+  var ${msgName} asyncapi.${pubConfig.message}
+  //constrcut your message here
```
Typo: construct

dxps reviewed (commented):
Just a heads-up for now (for the 1st one, I'll look into it myself as well, to see how harmful it can be):

- in `payloads.go`, there is this warning:
  ![image](https://user-images.githubusercontent.com/14140226/160246911-fa627f4a-f96b-4f01-95c4-1be61e21b8ca.png)
- in `main.go`, there is this warning:
   ![image](https://user-images.githubusercontent.com/14140226/160246978-ed42987c-f45a-4593-a2a3-a7d518c286b3.png)
   This can easily be solved by checking for that error, by introducing these after that call:
   ```go
	if err != nil {
		return err
	}
   ```

dxps commented:
Ok, I understand the 1st warning (`copylocks`) and it's quite problematic.

To solve it, the code should be slightly updated as follows:
1. In `payloads.go` file, `payloadToMessage` function should return a pointer to the message instead of a copy
  (that does a copy of an included Mutex, that's why the warning). Basically, it should look like this:
   ```go
   func PayloadToMessage(i interface{}) (*message.Message, error) {
        var m message.Message

        b, err := json.Marshal(i)
        if err != nil {
                return &m, nil
        }
        m.Payload = b

        return &m, nil
   }
   ```
2. And just adapt the handlers in `handlers.go` file, to call the `Publish(...)` method accordingly, that is:
    ```go
   return a.Publish("{topicName}", m)
   ```

dxps commented:
Last but not least, some small improvements that can be included in `main.go`:
1. Close the connection at the end.
    Although it's a generated code, it should guide others how to do it properly.
    Ofc, without it, the sample still works, and you could just see RabbitMQ's log complaining about it
    (with _client unexpectedly closed TCP connection_). The code to include right before the last `return nil` line:
    ```go
	if err := amqpPub.Close(); err != nil {
		log.Printf("Error while closing AMQP connection: %v\n", err)
	}
   ```
1. I might be too picky here, but `startAMQPPublishers(_)` function is not actually starting something (that should eventually be stopped), but actually it showcases how publishing should be done. Therefore, I'd rename it as `doAMQPPublishing()` or something like that. And ofc, the error message being logged in `main(_)` can be `error publishing: ...`.

derberg commented on package.json:
```diff
@@ -10,6 +10,7 @@
     "generate:readme:toc": "markdown-toc -i README.md",
     "generate:assets": "npm run generate:readme:toc",
     "bump:version": "npm --no-git-tag-version --allow-same-version version $VERSION",
+    "devTest": "ag test/asyncapi.yaml ./ -o output --force-write",
```
please use `:` notation for scripts name to be consistent, in this case `dev` would be enough tbh

derberg reviewed (commented):
I did not look into logic code as I see others already did it.

I only left one comment about one script name. Also please change PR title to `feat: add RabbitMQ publisher`. I also think it would be worth extending the readme overview with some more info on what users can generate with this template. Also good to add `development` section and explain the `dev` script.

One curiosity question: why did you add dependency related to babel?

anandsunderraman commented:
> here

@dxps again thanks for finding this issue I will look to address it

anandsunderraman commented:
> 1. Close the connection at the end.
>    Although it's a generated code, it should guide others how to do it properly.
>    Ofc, without it, the sample still works, and you could just see RabbitMQ's log complaining about it
>    (with _client unexpectedly closed TCP connection_). The code to include right before the last `return nil` line:

@dxps Completely agree with this suggestion. However I feel this needs more investigation to be done in the watermill.io fashion. I will create a separate issue to handle this

anandsunderraman commented:
> One curiosity question: why did you add dependency related to babel?
@derberg I introduced a change in this PR at: https://github.com/asyncapi/go-watermill-template/pull/137/files#diff-4bab0aa7cca57d1ac8ae38749f9f50f6c2c91b81eb46e28ad46ebcf0e375d1ceR41

```
const result = render(<SubscriptionHandlers channels={doc.channels()} />)
```

If I did not add babel I was getting the error:
```
Add @babel/preset-react (https://git.io/JfeDR) to the 'presets' section of your Babel config to enable transformation.
    If you want to leave it as-is, add @babel/plugin-syntax-jsx (https://git.io/vb4yA) to the 'plugins' section to enable parsing.
```

dxps commented:
Tested with the recently introduced updates.
LGTM :+1:

smoya commented on test/asyncapi.yaml:
```diff
@@ -1,4 +1,4 @@
-asyncapi: '2.1.0'
+asyncapi: '2.2.0'
```
Taking this opportunity to update it to the last (as per today) version: `2.3.0`.
```suggestion
asyncapi: '2.3.0'
```

smoya commented on test/files/docWithAMQPSubscriber.yml:
```diff
@@ -0,0 +1,83 @@
+asyncapi: '2.1.0'
```
Nitpick: Taking this opportunity to update it to the last (as per today) version: 2.3.0.

```suggestion
asyncapi: '2.3.0'
```

smoya replied:
Nitpick: Taking this opportunity to update it to the last (as per today) version: 2.3.0.
```suggestion
asyncapi: '2.3.0'
```

smoya reviewed (changes requested):
Pasting here as the conversation was marked as solved (now I unresolved).

https://github.com/asyncapi/go-watermill-template/pull/137/files#r847178655

> @anandsunderraman I reviewed the code again and I noticed you didn't add a guard checking if the operationId is present, forcing the user to set it or throwing an error otherwise. You have an example in [nodejs-ws-template](https://github.com/asyncapi/nodejs-ws-template/blob/c50a34556de766098e7b3b25d95fc064235e3728/filters/all.js#L21-L28).

Also, I would mention this in the README file as for example [java-spring-template](https://github.com/asyncapi/java-spring-template/blob/7feab3c2d5b2800828e25e28c30eb0deed56e90d/README.md#asyncapi-definitions) does.

anandsunderraman commented:
> Pasting here as the conversation was marked as solved (now I unresolved).
> 
> https://github.com/asyncapi/go-watermill-template/pull/137/files#r847178655
> 
> > @anandsunderraman I reviewed the code again and I noticed you didn't add a guard checking if the operationId is present, forcing the user to set it or throwing an error otherwise. You have an example in [nodejs-ws-template](https://github.com/asyncapi/nodejs-ws-template/blob/c50a34556de766098e7b3b25d95fc064235e3728/filters/all.js#L21-L28).
> 
> Also, I would mention this in the README file as for example [java-spring-template](https://github.com/asyncapi/java-spring-template/blob/7feab3c2d5b2800828e25e28c30eb0deed56e90d/README.md#asyncapi-definitions) does.

@smoya apologies. Thanks for pointing out to the other generator examples.
I have addressed them now and added unit tests.

smoya reviewed (approved):
Superb work @anandsunderraman! Thanks for such an effort; It has been a pleasure to review this PR :)

LGTM!🚀 🌔

anandsunderraman commented:
thanks @smoya !!
